### PR TITLE
Client Authentication - Bearer Body

### PR DESCRIPTION
**Pull request by anjuthomas**

### CODE REVIEW

File: src/clientAuth/bearerBody.js [hunk 1]
```diff
@@ -8,7 +14,6 @@ class BearerBody extends ClientAuthnMethod {
   }
 
   /**
-   * More complicated logic then I would have liked it to be
    *@param {*} cis Request class instance
```
Comment (TMSCH) [inline src/clientAuth/bearerBody.js]: You cannot use * param. be more specific (`Request`?). Same below.

File: test/clientAuth-bearerBody-test.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    require('../src/clientAuth/clientAuth').CLIENT_AUTHN_METHOD;
+const ResourceRequest = require('../oicMsg/oauth2/init.js').ResourceRequest;
+
+const CLIENT_ID = 'A';
+
+const CLIENT_CONF = {
+  'issuer': 'https://example.com/as',
+  'redirect_uris': ['https://example.com/cli/authz_cb'],
+  'client_secret': 'boarding pass',
+  'client_id': CLIENT_ID
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Add trailing comma

File: test/clientAuth-bearerBody-test.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+const REQ_ARGS = {
+  'redirect_uri': 'https://example.com/rp/cb',
+  'response_type': 'code'
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Add trailing comma

File: test/clientAuth-bearerBody-test.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+function getClient() {
+  let client = new Client();
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: `const` (you are not reassigning `client`)

File: test/clientAuth-bearerBody-test.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  client.init(CLIENT_AUTHN_METHOD, CLIENT_CONF);
+  let sdb = client.clientInfo.stateDb;
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Same here, `const` (and other occurrences in the file)

File: test/clientAuth-bearerBody-test.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+const BearerBody = require('../src/clientAuth/bearerBody').BearerBody;
+const Client = require('../src/oauth2/init').Client;
+const CLIENT_AUTHN_METHOD =
+    require('../src/clientAuth/clientAuth').CLIENT_AUTHN_METHOD;
+const ResourceRequest = require('../oicMsg/oauth2/init.js').ResourceRequest;
+
+const CLIENT_ID = 'A';
+
+const CLIENT_CONF = {
+  'issuer': 'https://example.com/as',
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: You don't need the quote around the keys here, remove here and below.

File: test/clientAuth-bearerBody-test.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  client.init(CLIENT_AUTHN_METHOD, CLIENT_CONF);
+  let sdb = client.clientInfo.stateDb;
+  sdb.dict = {};
+  sdb.dict['ABCDE'] = {'code': 'accessCode'};
+  client.clientInfo.clientSecret = 'boardingPass';
+  return client;
+}
+
+describe('Test bearer body', function() {
+  var client;
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: `let` (no `var` in ES6)

File: test/clientAuth-bearerBody-test.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  let client = new Client();
+  client.init(CLIENT_AUTHN_METHOD, CLIENT_CONF);
+  let sdb = client.clientInfo.stateDb;
+  sdb.dict = {};
+  sdb.dict['ABCDE'] = {'code': 'accessCode'};
+  client.clientInfo.clientSecret = 'boardingPass';
+  return client;
+}
+
+describe('Test bearer body', function() {
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Use arrow functions `() => {}` for anonymous functions (here and everywhere else below).

Note: you can declare functions the way you did above though for `getClient`.

File: test/clientAuth-bearerBody-test.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  var client;
+  beforeEach(function() {
+    client = getClient();
+  });
+  it('test construct with request args', function() {
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Add new line before.

**Review by TMSCH — CHANGES_REQUESTED** [review]

File: src/clientAuth/bearerBody.js (hunk 1 above)
Comment (anjuthomas) [reply src/clientAuth/bearerBody.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 2 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 3 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 4 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 5 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 6 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 7 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 8 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 9 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: src/clientAuth/bearerBody.js [hunk 10]
```diff
@@ -8,12 +14,11 @@ class BearerBody extends ClientAuthnMethod {
   }
 
   /**
-   * More complicated logic then I would have liked it to be
-   *@param {*} cis Request class instance
-   *@param {*} ci Client information
-   *@param {*} requestArgs Request arguments
-   *@param {*} httpArgs HTTP header arguments
-   *@param {*} kwargs
+   *@param {ResourceRequest} cis Request class instance
```
Comment (TMSCH) [inline src/clientAuth/bearerBody.js]: Add a space between * and @param (sorry I missed this earlier).

Comment (TMSCH) [inline src/clientAuth/bearerBody.js]: This is implicitly nullable and not recommended. Use this:

`@param {!ResourceRequest}` to make it non nullable,
`@param {?ResourceRequest}` to make it nullable.

Same goes for the other objects below. Only primitives (`string`, `boolean`) are non null by default.

File: src/clientAuth/bearerBody.js [hunk 11]
```diff
@@ -8,12 +14,11 @@ class BearerBody extends ClientAuthnMethod {
...
+   *@param {ClientInfo} cliInfo Client information
+   *@param {Object.<string, string>} requestArgs Request arguments
```
Comment (TMSCH) [inline src/clientAuth/bearerBody.js]: `{!Object<string, string>}`

Even better: provide the interface of the object here. No need to do in this CL though.

File: test/clientAuth-bearerBody-test.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  redirect_uri: 'https://example.com/rp/cb',
+  response_type: 'code',
+};
+
+function getClient() {
+  const client = new Client();
+  client.init(CLIENT_AUTHN_METHOD, CLIENT_CONF);
+  const sdb = client.clientInfo.stateDb;
+  sdb.dict = {};
+  sdb.dict['ABCDE'] = {'code': 'accessCode'};
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: You can also remove the quotes in the keys here:

`sdb.dict['ABCDE'] = {code: 'accessCode'};`

File: test/clientAuth-bearerBody-test.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    const sdb = client.clientInfo.stateDb;
+    sdb['FFFFF'] = {};
+    const resp = new AuthorizationResponse('code', 'FFFFF')
+    sdb.addResponse(resp);
+    const atr = new AccessTokenResponse({
+      'access_token': '2YotnFZFEjr1zCsicMWpAA',
+      'token_type': 'example',
+      'refresh_token': 'tGzv3JOkF0XG5Qx2TlKWIA',
+      'example_parameter': 'example_value',
+      'scope': ['inner', 'outer']
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Add trailing comma here.

File: test/clientAuth-bearerBody-test.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    assert.deepEqual(httpArgs, undefined);
+  });
+
+  it('test construct with state', () => {
+    const sdb = client.clientInfo.stateDb;
+    sdb['FFFFF'] = {};
+    const resp = new AuthorizationResponse('code', 'FFFFF')
+    sdb.addResponse(resp);
+    const atr = new AccessTokenResponse({
+      'access_token': '2YotnFZFEjr1zCsicMWpAA',
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Remove quotes in keys here.

File: test/clientAuth-bearerBody-test.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+      'example_parameter': 'example_value',
+      'scope': ['inner', 'outer']
+    });
+    sdb.addResponse(atr, {'state': 'FFFFF'});
+    let cis = new ResourceRequest();
+    const list = new BearerBody().construct(
+        cis, client.clientInfo, {}, null, {'state': 'FFFFF'}, 'inner');
+    const httpArgs = list[0];
+    cis = list[1];
+    assert.deepEqual(cis['access_token'], '2YotnFZFEjr1zCsicMWpAA');
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: cis.access_token

File: test/clientAuth-bearerBody-test.js [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+      'access_token': '2YotnFZFEjr1zCsicMWpAA',
+      'token_type': 'example',
+      'refresh_token': 'tGzv3JOkF0XG5Qx2TlKWIA',
+      'example_parameter': 'example_value',
+      'scope': ['inner', 'outer']
+    });
+    sdb.addResponse(atr, {'state': 'FFFFF'});
+    let cis = new ResourceRequest();
+    const list = new BearerBody().construct(
+        cis, client.clientInfo, {}, null, {'state': 'FFFFF'}, 'inner');
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: {state: 'FFFFF'}

File: test/clientAuth-bearerBody-test.js [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+  it('test construct with request', () => {
+    const sdb = client.clientInfo.stateDb;
+    sdb['EEEE'] = {};
+    const resp = new AuthorizationResponse('auth_grant', 'EEEE');
+    new client.service['Authorization']().parseResponse(
+        resp, client.clientInfo, 'urlencoded');
+
+    const resp2 = new AccessTokenResponse({
+      'access_token': 'token1',
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js]: Remove keys quotes

**Review by TMSCH — CHANGES_REQUESTED** [review]

File: test/clientAuth-bearerBody-test.js (hunk 17 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 16 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 15 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 14 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 13 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: test/clientAuth-bearerBody-test.js (hunk 12 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Fixed

File: src/clientAuth/bearerBody.js (hunk 10 above)
Comment (anjuthomas) [reply src/clientAuth/bearerBody.js]: Fixed

Comment (anjuthomas) [reply src/clientAuth/bearerBody.js]: Fixed

File: src/clientAuth/bearerBody.js [hunk 18]
```diff
@@ -23,24 +28,24 @@ class BearerBody extends ClientAuthnMethod {
     if (Object.keys(cis).indexOf('access_token') !== -1) {
```
Comment (TMSCH) [inline src/clientAuth/bearerBody.js]: You can just do:

`if (!cis.access_token)`

File: test/clientAuth-bearerBody-test.js [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    cis = list[1];
+
+    assert.deepEqual(cis.access_token, 'Sesame');
+    assert.deepEqual(httpArgs, undefined);
+  });
+
+  it('test construct with state', () => {
+    const sdb = client.clientInfo.stateDb;
+    sdb.FFFFF = {};
+    const resp = new AuthorizationResponse('code', 'FFFFF')
```
Comment (TMSCH) [inline test/clientAuth-bearerBody-test.js:60]: Here you're dynamically accessing the property `FFFFF` of sdb. In that case, sdb['FFFFF'] has to be quoted. It's not such an issue though as these are tests so they're not compiled. You can keep it the way it is.

**Review by TMSCH — CHANGES_REQUESTED** [review]

File: src/clientAuth/bearerBody.js (hunk 18 above)
Comment (anjuthomas) [reply src/clientAuth/bearerBody.js]: Fixed. Should be cis.access_token instead of !cis.access_token

File: src/clientAuth/bearerBody.js [hunk 20]
```diff
@@ -8,39 +14,38 @@ class BearerBody extends ClientAuthnMethod {
...
    */
   construct(cis, cliInfo, requestArgs, httpArgs, kwargs) {
     if (requestArgs === null) {
       requestArgs = {};
     }
 
-    if (Object.keys(cis).indexOf('access_token') !== -1) {
+    if (cis.access_token) {
       return;
     } else {
```
Comment (TMSCH) [inline src/clientAuth/bearerBody.js]: No need to use `else` here as you return just above.

**Review by TMSCH — APPROVED** [review]
Just one small change and you can merge.

File: test/clientAuth-bearerBody-test.js (hunk 19 above)
Comment (anjuthomas) [reply test/clientAuth-bearerBody-test.js]: Got it

**Pull request merged**